### PR TITLE
Add additional F5 sensor support

### PR DESCRIPTION
**Pull request by boudreau**
#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you signed the [Contributors agreement](http://docs.librenms.org/General/Contributing/)
- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

### CODE REVIEW

File: includes/discovery/sensors/states/f5.inc.php [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+                    'state_descr' => $value[1],
+                    'state_draw_graph' => $value[2],
+                    'state_value' => $value[3],         // Value polled from device
+                    'state_generic_value' => $value[4]  // Set value based on the Nagios standard 0=OK, 1=Warning, 2=Critical, 3=Unknown
+                );
+                dbInsert($insert, 'state_translations');
+            }
+        }
+
+      foreach (array_keys($temp) as $index) {
```
Comment (murrant) [inline includes/discovery/sensors/states/f5.inc.php]: You can do foreach($temp as $index => $data) here.

File: includes/discovery/sensors/states/f5.inc.php [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+          $descr           = "sysChassisPowerSupplyStatus.".$temp[$index]['sysChassisPowerSupplyIndex'];
+          $current         = $temp[$index]['sysChassisPowerSupplyStatus'];
```
Comment (murrant) [inline includes/discovery/sensors/states/f5.inc.php]: and $current = $data['sysChassisPowerSupplyStatus']; here

File: includes/discovery/sensors/states/f5.inc.php [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+          $sensorType      = 'f5';
+          $oid             = '1.3.6.1.4.1.3375.2.1.3.2.2.2.1.2.'.$index;
+          $oid_status	 = snmp_get($device, $oid, '-Oqv');
```
Comment (murrant) [inline includes/discovery/sensors/states/f5.inc.php]: Remove this line, it is unused and causes many extra snmp_get() calls

**Comment by laf** [comment]
Also lots of formatting issues, please see our docs about running pre-commit checks.

File: includes/defaults.inc.php [hunk 4]
```diff
@@ -703,7 +703,9 @@
 $config['poller_modules']['cisco-mac-accounting']        = 1;
 $config['poller_modules']['cipsec-tunnels']              = 1;
 $config['poller_modules']['cisco-ace-loadbalancer']      = 1;
+$config['poller_modules']['F5-loadbalancer']             = 1;
```
Comment (laf) [inline includes/defaults.inc.php]: Not sure you've added these in. They aren't poller modules.

File: includes/defaults.inc.php [hunk 5]
```diff
@@ -703,7 +703,9 @@
...
 $config['poller_modules']['cisco-ace-serverfarms']       = 1;
+$config['poller_modules']['F5-serverfarms']              = 1;
```
Comment (laf) [inline includes/defaults.inc.php]: Same goes for this.

File: includes/discovery/sensors/fanspeeds/f5.inc.php [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    $type    = 'f5';
+
+    foreach (explode("\n", $oids) as $data) {
+        $data = trim($data);
+        if ($data) {
+            list($oid,$descr) = explode(' ', $data, 2);
+            $split_oid        = explode('.', $oid);
+            $index            = $split_oid[(count($split_oid) - 1)];
+            $descr            = 'Fan Speed '.$index;
+            $oid              = '1.3.6.1.4.1.3375.2.1.3.2.1.2.1.3.'.$index;
```
Comment (laf) [inline includes/discovery/sensors/fanspeeds/f5.inc.php]: This should start with a . - .1.3.6

File: includes/discovery/sensors/states/f5.inc.php [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+                );
+                dbInsert($insert, 'state_translations');
+            }
+        }
+
+        foreach ($temp as $index => $data) {
+            $descr           = "sysChassisPowerSupplyStatus.".$temp[$index]['sysChassisPowerSupplyIndex'];
+            $current         = $data['sysChassisPowerSupplyStatus'];
+            $sensorType      = 'f5';
+            $oid             = '1.3.6.1.4.1.3375.2.1.3.2.2.2.1.2.'.$index;
```
Comment (laf) [inline includes/discovery/sensors/states/f5.inc.php]: This should start with a . - .1.3.6

File: includes/discovery/sensors/states/f5.inc.php [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+                );
+                dbInsert($insert, 'state_translations');
+            }
+        }
+
+        foreach (array_keys($temp1) as $index) {
+            $descr           = "sysCmFailoverStatusId.".$temp1[$index]['sysCmFailoverStatusId'];
+            $current         = $temp1[$index]['sysCmFailoverStatusId'];
+            $sensorType      = 'f5';
+            $oid             = '1.3.6.1.4.1.3375.2.1.14.3.1.'.$index;
```
Comment (laf) [inline includes/discovery/sensors/states/f5.inc.php]: This should start with a . - .1.3.6

File: includes/discovery/sensors/states/f5.inc.php [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+                );
+                dbInsert($insert, 'state_translations');
+            }
+        }
+
+        foreach (array_keys($temp1) as $index) {
+            $descr           = "sysCmFailoverStatusColor.".$index;
+            $current         = $temp1[$index]['sysCmFailoverStatusColor'];
+            $sensorType      = 'f5';
+            $oid             = '1.3.6.1.4.1.3375.2.1.14.3.3.'.$index;
```
Comment (laf) [inline includes/discovery/sensors/states/f5.inc.php]: This should start with a . - .1.3.6

File: includes/discovery/sensors/states/f5.inc.php [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+                );
+                dbInsert($insert, 'state_translations');
+            }
+        }
+
+        foreach (array_keys($temp3) as $index) {
+            $descr           = "Fan Speed Status ".$index;
+            $current         = $temp3[$index]['sysChassisFanStatus'];
+            $sensorType      = 'f5';
+            $oid             = '1.3.6.1.4.1.3375.2.1.3.2.1.2.1.2.'.$index;
```
Comment (laf) [inline includes/discovery/sensors/states/f5.inc.php]: This should start with a . - .1.3.6

**Comment by boudreau** [comment]
Hi,

Sorry, got mixed up with another project, no changes should be applied on the includes/defaults.inc.php<https://github.com/librenms/librenms/pull/4642#pullrequestreview-2036461>: file.

I added the missing ‘.’ on the OIDs.

Thanks

Yves

**Comment by Rosiak** [comment]
Sorry to bother again, but would it be possible to clean up the state sensor code?
You should be able to use includes/discovery/sensors/states/cisco.inc.php as an example.

File: includes/discovery/sensors/fanspeeds/f5.inc.php [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+    foreach (explode("\n", $oids) as $data) {
+        $data = trim($data);
+        if ($data) {
+            list($oid,$descr) = explode(' ', $data, 2);
+            $split_oid        = explode('.', $oid);
+            $index            = $split_oid[(count($split_oid) - 1)];
+            $descr            = 'Fan Speed '.$index;
+            $oid              = '.1.3.6.1.4.1.3375.2.1.3.2.1.2.1.3.'.$index;
+            $fanspeed         = (snmp_get($device, $oid, '-Oqv') / $divisor);
```
Comment (laf) [inline includes/discovery/sensors/fanspeeds/f5.inc.php]: Haven't we got fanspeed from the sysChassisFanSpeed walk?

File: includes/discovery/sensors/states/f5.inc.php [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+                dbInsert($insert, 'state_translations');
+            }
+        }
+
+        foreach (array_keys($temp1) as $index) {
+            $descr           = "sysCmFailoverStatusColor.".$index;
+            $current         = $temp1[$index]['sysCmFailoverStatusColor'];
+            $sensorType      = 'f5';
+            $oid             = '.1.3.6.1.4.1.3375.2.1.14.3.3.'.$index;
+            $oid_status    = snmp_get($device, $oid, '-Oqev');
```
Comment (laf) [inline includes/discovery/sensors/states/f5.inc.php]: Don't we have that from the snmpwalk of sysCmFailoverStatusColor

File: includes/discovery/sensors/states/f5.inc.php [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+                );
+                dbInsert($insert, 'state_translations');
+            }
+        }
+
+        foreach (array_keys($temp1) as $index) {
+            $descr           = "sysCmFailoverStatusId.".$temp1[$index]['sysCmFailoverStatusId'];
+            $current         = $temp1[$index]['sysCmFailoverStatusId'];
+            $sensorType      = 'f5';
+            $oid             = '.1.3.6.1.4.1.3375.2.1.14.3.1.'.$index;
```
Comment (laf) [inline includes/discovery/sensors/states/f5.inc.php]: Again don't we have this value from the walk already?

File: includes/discovery/sensors/states/f5.inc.php [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+            }
+        }
+
+        foreach (array_keys($temp3) as $index) {
+            $descr           = "Fan Speed Status ".$index;
+            $current         = $temp3[$index]['sysChassisFanStatus'];
+            $sensorType      = 'f5';
+            $oid             = '.1.3.6.1.4.1.3375.2.1.3.2.1.2.1.2.'.$index;
+            $oid_status    = snmp_get($device, $oid, '-Oqv');
+            discover_sensor($valid['sensor'], 'state', $device, $oid, $index, $state_name, $descr, '1', '1', null, null, null, null, $current, 'snmp', $index);
```
Comment (laf) [inline includes/discovery/sensors/states/f5.inc.php]: Don't we have this value already from the snmpwalk?

File: includes/discovery/sensors/temperatures/comware.inc.php [hunk 15]
```diff
@@ -13,9 +13,11 @@
 if ($device['os'] == 'comware') {
     echo 'Comware ';
     
-    $entphydata = dbFetchRows("SELECT `entPhysicalIndex`, `entPhysicalClass`, `entPhysicalName` FROM `entPhysical` WHERE `device_id` = ? AND `entPhysicalClass` REGEXP 'module|sensor' ORDER BY `entPhysicalIndex`", array(
-        $device['device_id']
-    ));
+    $entphydata = dbFetchRows(
```
Comment (laf) [inline includes/discovery/sensors/temperatures/comware.inc.php]: The formatting of this seems unnecessary.

File: includes/discovery/sensors/temperatures/comware.inc.php [hunk 16]
```diff
@@ -36,10 +38,12 @@
     foreach ($comware_oids as $index => $entry) {
         if (is_numeric($entry['hh3cTransceiverTemperature']) && $entry['hh3cTransceiverTemperature'] != 2147483647) {
             $oid                       = '.1.3.6.1.4.1.25506.2.70.1.1.1.15.' . $index;
-            $dbquery                   = dbFetchRows("SELECT `ifDescr` FROM `ports` WHERE `ifIndex`= ? AND `device_id` = ? AND `ifAdminStatus` = 'up'", array(
-                $index,
+            $dbquery                   = dbFetchRows(
```
Comment (laf) [inline includes/discovery/sensors/temperatures/comware.inc.php]: The formatting of this seems unnecessary.

File: includes/discovery/sensors/temperatures/f5.inc.php [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+
+    if (is_array($f5)) {
+        echo "sysChassisTempTable ";
+
+        foreach (array_keys($f5) as $index) {
+            $descr           = "sysChassisTempTemperature.".$f5[$index]['sysChassisTempIndex'];
+            $current         = $f5[$index]['sysChassisTempTemperature'];
+            $sensorType      = 'f5';
+            $oid             = '.1.3.6.1.4.1.3375.2.1.3.2.3.2.1.2.'.$index;
+            $low_limit       =  null;
```
Comment (laf) [inline includes/discovery/sensors/temperatures/f5.inc.php]: This seems unnecessary to set all these variables when they are only ever null, just pass null to the function.

File: includes/discovery/sensors/temperatures/f5.inc.php [hunk 18]
```diff
@@ -0,0 +1,52 @@
+<?php
+
+if ($device['os'] == 'f5') {
+    $f5 = null;
```
Comment (laf) [inline includes/discovery/sensors/temperatures/f5.inc.php]: should be `$f5 = array()`` or better yet `$f5_chassis = array();`

File: includes/discovery/sensors/temperatures/f5.inc.php [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+            $low_warn_limit  =  null;
+            $high_warn_limit =  null;
+            $high_limit      =  null;
+
+            discover_sensor($valid['sensor'], 'temperature', $device, $oid, $index, $sensorType, $descr, '1', '1', $low_limit, $low_warn_limit, $high_warn_limit, $high_limit, $current);
+        }
+    }
+
+    // Get the CPU Temperature values
+    $f5cpu = null;
```
Comment (laf) [inline includes/discovery/sensors/temperatures/f5.inc.php]: Should be `$f5_cpu = array();`

File: includes/discovery/sensors/temperatures/f5.inc.php [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+    //    $f5cpu = snmpwalk_cache_oid($device, 'sysCpuSensorTemperature', array(), 'F5-BIGIP-SYSTEM-MIB');
+    $f5cpu = snmpwalk_cache_multi_oid($device, 'sysCpuSensorTemperature', array(), 'F5-BIGIP-SYSTEM-MIB');
```
Comment (laf) [inline includes/discovery/sensors/temperatures/f5.inc.php]: Seems more appropriate to walk ` sysCpuSensorEntry` so you don't need the other two snmp_get()'s

**Comment by laf** [comment]
I've put some comments in this. Please also remove the .txt extension from all MIBs.

**Comment by laf** [comment]
bump @boudreau

File: includes/discovery/sensors/fanspeeds/f5.inc.php (hunk 11 above)
Comment (boudreau) [inline includes/discovery/sensors/fanspeeds/f5.inc.php]: Not sure, I'm following?
I did the same code structure as in dell.inc.php in includes/discovery/sensors/fanspeeds. 
It also looks the same in the lmsensors.inc.php. They snmp_walk and then it snmp_get ?

**Pull request merged**